src/main/java/picard/sam/StabilizeQualityScores.java

```diff
@@ -0,0 +1,383 @@
+
+    static final String USAGE_SUMMARY = "Stabilize quality scores.";
+    static final String USAGE_DETAILS = "This tool bins and stabilizes the quality scores of a SAM or BAM file " +
+            "to eliminate short runs of zigzaggy qualities.\n" +
+            "Binning strategies available:\n\n" +
+            "THRESHOLD N:\n " + BIN_THRESHOLD_STRATEGY_DESCRIPTION + "\n" +
+            "BINS Q1 Q2 Q3 ...:\n " + BIN_BINS_STRATEGY_DESCRIPTION + "\n" +
+
+            "\n" +
+
+            "Stabilization strategies available:\n\n" +
+            "DROP N:\n " + STABILIZE_DROP_STRATEGY_DESCRIPTION + "\n" +
+            "AVERAGE N:\n " + STABILIZE_AVERAGE_STRATEGY_DESCRIPTION + "\n" +
+            "MERGE N:\n " + STABILIZE_MERGE_STRATEGY_DESCRIPTION + "\n" +
+
+            "<h3>Usage examples:</h3>\n" +
+            "<h4>Stabilize file, using a Q30 threshold and dropping runs of oscillations of length &lt; 4</h4>\n"+
+            "\n"+
+            "java -jar picard.jar StabilizeQualityScores \\\n" +
+            "      I=input.bam \\\n" +
+            "      O=stabilized.bam \\\n" +
+            "      THRESHOLD=30\n"+
+            "      DROP=4\n";
+
+    @Argument(shortName = StandardOptionDefinitions.INPUT_SHORT_NAME, doc = "The input SAM or BAM file to stabilize.")
+    public File INPUT;
+
+    @Argument(shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME, doc = "The stabilized SAM or BAM file to write.")
+    public File OUTPUT;
+
+    @Argument(shortName = "BANKS", optional = true, doc = "For use with THRESHOLD. Sets the above-threshold value.")
+    public Integer THRESHOLD_UP = null;
+
+    @Argument(shortName = "T", doc = "Any qual <T becomes Q2. Any qual >=T becomes T, or the value of THRESHOLD_UP if specified.", mutex = {"BINS", "CUTOFF"})
+    public Integer THRESHOLD = 20;
+
+    @Argument(shortName = "B", doc = "All qualities round to nearest bin in this list.", mutex = {"THRESHOLD", "CUTOFF"})
+    public List<Integer> BINS = null;
+
+    @Argument(shortName = "C", doc = "Find the cutoff where the end of the read drops consistently at or below this qual. Set the end of the read to Q2 and the rest to Q-average.", mutex = {"THRESHOLD", "BINS", "DROP", "AVERAGE", "MERGE"})
+    public Integer CUTOFF = null;
+
+    @Argument(shortName = "D", doc = "Drop runs of <= N quals to the lowest qual in the run.", mutex = {"AVERAGE", "MERGE", "CUTOFF"})
+    public Integer DROP = null;
+
+    @Argument(shortName = "A", doc = "Average runs of <= N quals to the binned average quality of the run.", mutex = {"DROP", "MERGE", "CUTOFF"})
+    public Integer AVERAGE = 4;
+
+    @Argument(shortName = "M", doc = "Merge runs of <= N quals to the qual immediately previous to the run.", mutex = {"AVERAGE", "DROP", "CUTOFF"})
+    public Integer MERGE = null;
+
+    abstract static class Binner {
+        public abstract int[] bin(int[] quals);
+    }
+
+    static class ThresholdBinner extends Binner {
+        protected Integer threshold;
```

Comment:
In general, only declare boxed if they can contain null; use `int` etc otherwise. It looks like these cannot contain null.